--- elinks-0.11.1/src/session/task.c.kzak	2006-10-11 09:26:28.000000000 +0200
+++ elinks-0.11.1/src/session/task.c	2006-10-11 09:27:29.000000000 +0200
@@ -538,7 +538,8 @@
 	}
 
 	if (is_in_result_state(download->state) && download->state != S_OK) {
-		print_error_dialog(ses, download->state, download->conn->uri,
+		print_error_dialog(ses, download->state, 
+				   download->conn ? download->conn->uri : NULL,
 				   download->pri);
 		if (d == DO_MOVE_ABORT) reload(ses, CACHE_MODE_NORMAL);
 	}
